### PR TITLE
LDAP-53 Add support for StartTLS

### PR DESCRIPTION
**Pull request by Godin**
https://jira.sonarsource.com/browse/LDAP-53

### CODE REVIEW

**Comment by n-b-sonar** [comment]
Tested with SQ 5.6 + [osixia/docker-openldap](https://github.com/osixia/docker-openldap) (together with [osixia/docker-phpLDAPadmin](https://github.com/osixia/docker-phpLDAPadmin)):
- first connection to LDAP server during SonarQube startup is successful
- all subsequent connections (whenever a user attempts to log in SonarQube) fail with:

> javax.naming.NamingException: [LDAP: error code 1 - TLS already started]

Logs (`com.sun.jndi.ldap.connect.pool=true`) and [code](https://github.com/SonarSource/sonar-ldap/blob/299e23e40fc07b5425c9c05458abae4bf8c9c7c7/src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java#L88) show that LDAP connections are pooled, so my guess is that this patch is missing logic to check whether TLS is already started on the re-used context.

Full logs attached (very verbose as _javax_ ssl logs were also enabled): [sonar_starttls.txt](https://github.com/SonarSource/sonar-ldap/files/304928/sonar_starttls.txt)

**Comment by Godin** [comment]
Note to myself: pooling is enabled for `createBindContext`, which is used to test connection and for searches.

File: src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java [hunk 1]
```diff
@@ -97,8 +97,10 @@ public InitialDirContext createUserContext(String principal, String credentials)
   }
 
   private InitialDirContext createInitialDirContext(String principal, String credentials, boolean pooling) throws NamingException {
-    InitialLdapContext ctx = new InitialLdapContext(getEnvironment(principal, credentials, pooling), null);
+    final InitialLdapContext ctx;
     if (startTLS) {
+      ctx = new InitialLdapContext(getEnvironment(null, null, /* TODO(Godin): connection pooling requires proper TLS shutdown */false), null);
```
Comment (Godin) [inline src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java]: Note for myself - http://docs.oracle.com/javase/7/docs/technotes/guides/jndi/jndi-ldap.html#POOL

File: sonar-ldap-plugin/src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java [hunk 2]
```diff
@@ -92,7 +97,25 @@ public InitialDirContext createUserContext(String principal, String credentials)
   }
 
   private InitialDirContext createInitialDirContext(String principal, String credentials, boolean pooling) throws NamingException {
-    return new InitialLdapContext(getEnvironment(principal, credentials, pooling), null);
+    final InitialLdapContext ctx;
+    if (startTLS) {
+      ctx = new InitialLdapContext(getEnvironment(null, null, /* TODO(Godin): connection pooling requires proper TLS shutdown */false), null);
```
Comment (SonarTech) [inline sonar-ldap-plugin/src/main/java/org/sonar/plugins/ldap/LdapContextFactory.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* INFO 1 info

Watch the comments in this conversation to review them.

**Pull request merged**